### ISSUE TITLE
Multi-task learning

### ISSUE THREAD

**Issue by dirkgr** [allenai/allennlp#4422]
# Goals

* Be able to train VilBERT as in https://arxiv.org/abs/1912.02315 (12-in-1: Multi-Task Vision and Language Representation Learning)
* Be able to do a similar thing for NLP, i.e., fine-tune BERT for 12 tasks the same way

# Data Readers

We’ll write a `MultitaskReader` that takes a `Dict[str, DatasetReader]` as one of its `__init__()` arguments. This way we define the different datasets we want to read. Much like the `InterleavingDatasetReader`, this reader’s `_read()` method takes a `Dict[str, str]`, so that we can pass the right path to the right inner reader. Accordingly, the `*_data_path` keys in the config need to be dictionaries.

`MultitaskReader`’s `read()` method will return a `MultitaskDataset`. `MultitaskDataset` will act like `ChainDataset` (chaining the inner datasets into one large one), with two differences:
 * It adds an extra “task” field to each instance. This field is a `MetadataField` that contains the name of the task.
 * It gives access directly to the inner readers with two functions:
    * `tasks() -> List[str]`
    * `reader_for_task(task: str) -> DatasetReader`

# Data Loaders

We’ll write a `MultitaskDataLoader` that takes a `Dict[str, DataLoader]` as a `__init__()` argument that defines the different data loaders for each task. It also takes a `MultitaskScheduler` that determines which task gets batched next. The `__iter__()` method batches instances from the next task and returns them as a `TensorDict`.

When it does this, it makes one change to the `TensorDicts` that come from the inner data loaders: The instances will have a “task” field identifying the task, so the batch with those instances will contain a `List[str]` with the task in it. Since we know that one batch always comes from a single task, we replace that field with a single `str` containing the task name.

Note that this design makes it possible to define different batch sizes and shuffling options per task.

## Multitask Schedulers
Multitask schedulers are the only type of object that is genuinely new. As usual, we should have a base class with a a few implementations:
 * `“roundrobin”`: This should be the default scheduler that you get if you don’t specify anything. Optionally, this should take counts or a schedule, in case you want to do something more complicated. Counts are formatted as `Dict[str, int]`, specifying how often to repeat a given task. A schedule is a `List[str]`, for example `[“vqa”, “nlvr2”, “vqa”]`, to get that specific sequence of tasks.
 * `“stochastic”`, where you get a random task every time. Optionally you can specify weights for each task.

# Models
We’ll write a `MultitaskModel` base class. Derivations of this class hold some parameters within themselves (the “backbone”), and also support one or more “heads”, which perform the actual tasks. Every head is a `Model` in its own right, i.e., it returns a loss from the `forward()` method.

The `MultitaskModel`’s `__init__()` method must take a `Dict[str, Model]` that describes the task-specific heads, each with a name. Optionally, it should take a `Dict[str, float]`, containing scaling factors for the loss for each task.

The `forward()` method should take the following arguments:
 * `task: str`: This is a field that identifies the head to run. It’s created by the `MultitaskDataLoader`.
 * Some fields that are common to every task. The `MultitaskModel` runs these fields through the backbone to produce inputs for the heads.
 * `**kwargs`, which are forwarded to the heads unchanged.

The heads return a `dict` of results, including a loss. The `MultitaskModel` renames all of the results such that they are prefixed by the head name. For example, if the `“vqa”` head returns a tensor under the name `“logits”`, the `MultitaskModel` returns a tensor under the name `“vqa_logits”`. In addition, it computes a loss based on the loss from the head, and the loss weight for the task.

Similarly, the `MultitaskModel`’s `get_metrics()` call will call `get_metrics()` on all of the heads, and prefix the results with the head’s names.

**Comment by matt-gardner** [comment]
My first pass at how I would design something for this:

You have a `TextEncoder` or a `TextAndImageEncoder`, which goes from inputs to an encoded representation that's ready to have various task heads applied.  (Though, really, the `TextFieldEmbedder` already does what the `TextEncoder` would do, so we may not need anything new there, just for text+images.)  Then you have a `MultiTaskTextAndImageModel` that takes as `__init__` arguments a `TextAndImageEncoder` and a list of prediction heads.  `forward` is a bit tricky, you probably have to have it accept `**kwargs` and make assumptions about the names of things that it gets.  But assuming you can work that out, you then have the model apply whatever heads are required based on the inputs it receives, compute a joint loss, and that's it.  You can configure this pretty easily to add another head just by adding another prediction head to the list, and an appropriate dataset inside a multi-task dataset reader.

**Comment by jiasenlu** [comment]
I think what @matt-gardner suggested about model forward pass makes sense. For multi-task dataloader, we can borrow the existing code from M3Transofrmers. What we need is to call `dataloader.set_task(i)` before each iteration, and it will output the batched data for different tasks.

**Comment by dirkgr** [comment]
I updated this issue's description to be basically a spec, ready to implement. @wlhgtc, does it address your concerns?

**Comment by dirkgr** [comment]
I put the description of Stop&Go into #4566, since we can do that separately from this.

**Comment by wlhgtc** [comment]
> I updated this issue's description to be basically a spec, ready to implement. @wlhgtc, does it address your concerns?

Sure.

**Comment by schmmd** [comment]
@dirkgr  is going to check if there's anything left to do.

**Comment by mateuszpieniak** [comment]
How can we distribute a single instance to two or more heads?

**Comment by jbrry** [comment]
@mateuszpieniak A quick workaround I ended up doing was introducing a boolean value `multiple_heads_one_data_source` for the `multitask` model, where if specified, skips some of the dataset reader to task-head mapping logic.

It's not properly tested but served me for my needs. You can see some of the control flow [here](https://github.com/jbrry/IWPT-2021-shared-task/blob/cbabdb787ad186f64f5b9461b5264e948245972d/multitask_parser/models/multitask_v2.py#L138).

**Comment by dirkgr** [comment]
> How can we distribute a single instance to two or more heads?

To support this properly I would write another version of `MultitaskModel` that supports this. `MultitaskModel` is not that big a class anyways. But if you have a workaround, that's good too.
